DuckDuckGo/Subscription/PrivacyProDataReporter.swift

```diff
@@ -0,0 +1,289 @@
+//
+//  PrivacyProDataReporter.swift
+//  DuckDuckGo
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Core
+import BrowserServicesKit
+import DDGSync
+
+enum PrivacyProPromoParameters: String, CaseIterable {
+    case isReinstall
+    case fireButtonUser = "fireButtonUsed"
+    case syncUsed
+    case fireproofingUsed
+    case appOnboardingCompleted
+    case emailEnabled
+    case widgetAdded
+    case frequentUser
+    case longTermUser
+    case autofillUser
+    case validOpenTabsCount
+    case searchUser
+
+    /// Pick a randomized subset of parameters each time they are attached to a pixel
+    static func randomizedSubset(excluding excludedParameters: [PrivacyProPromoParameters] = []) -> [PrivacyProPromoParameters] {
+        let allParameters = Set(allCases).subtracting(Set(excludedParameters))
+        return Array(allParameters.shuffled().prefix(Int(allCases.count * 2/3)))
+    }
+}
+
+// swiftlint:disable identifier_name
+final class PrivacyProDataReporter {
+    enum Key {
+        static let fireCountKey = "com.duckduckgo.ios.privacypropromo.FireCount"
+        static let isWidgetAddedKey = "com.duckduckgo.ios.privacypropromo.WidgetAdded"
+        static let applicationLastSessionEndedKey = "com.duckduckgo.ios.privacypropromo.ApplicationLastSessionEnded"
+        static let searchCountKey = "com.duckduckgo.ios.privacypropromo.SearchCount"
+    }
+
+    enum Constants {
+        static let includedOrigins = "origins"
+    }
+
+    enum UseCase {
+        case messageID(String)
+        case origin(String?)
+        case debug
+    }
+
+    public static let shared = PrivacyProDataReporter()
```

Comment:
Let's replace the singleton with an instance of this class that's stored on the app delegate if possible - we're aiming to introduce no new singletons given their harm on testability. There are places in this PR that the singleton usage can impact unit testing, so I think it's worth resolving.